lib/src/requests.dart

```
@@ -83,22 +107,35 @@ class Requests {
     return uri.host;
   }
 
-  static Future<dynamic> _handleHttpResponse(String url, String hostname, http.Response response, bool json, bool persistCookies) async {
+  static Future<dynamic> _handleHttpResponse(
+      String url,
+      String hostname,
+      http.Response response,
+      bool json,
+      bool persistCookies,
+      shouldThrow,
+      dataOnly) async {
     int statusCode = response.statusCode;
-    dynamic responseBody = "";
+    bool hasClientOrServerError = (400 <= statusCode) && (statusCode < 600);
+    dynamic responseContent = "";
     dynamic parseException;
 
     try {
-      responseBody = utf8.decode(response.bodyBytes, allowMalformed: true);
+      responseContent = utf8.decode(response.bodyBytes, allowMalformed: true);
     } catch (e) {
       parseException = e;
     }
 
-    bool hasError = (400 <= statusCode) && (statusCode < 600);
-    if (hasError) {
-      var errorEvent = {"statusCode": statusCode, "url": url, "response": responseBody};
+    if (json) responseContent = Common.fromJson(responseContent);
+
+    if (hasClientOrServerError && shouldThrow) {
+      var errorEvent = {
+        "statusCode": statusCode,
+        "url": url,
+        "response": responseContent
+      };
       onError.publish(errorEvent);
-      throw Exception("Invalid HTTP status code $statusCode for url $url");
+      throw Response(responseContent, statusCode, !hasClientOrServerError);
```

I prefer to throw a custom exception with all data in his instance
```
throw HTTPException(message, statusCode, responseContent)
```